padertorch/train/trainer.py

```diff
@@ -266,48 +276,111 @@ def train(
+
+                    if new_epoch:
+                        new_epoch = False
+                    elif minibatch_index == 0:
+                        # Call pre_step after getting the next example,
+                        # to correctly detect the next epoch
                         for hook in hooks:
                             hook.pre_step(self)
-                    with self.train_timer['time_per_step']:
-                        model_output, review = self.train_step(
-                            example,
-                            optimize=(self.iteration+1) % self.virtual_minibatch_size == 0,
-                        )
 
-                    for hook in hooks:
-                        hook.post_step(self, example, model_output, review)
+                    if len(device) == 1:
+                        with self.train_timer['time_per_step']:
+                            example, = example
+                            loss, example, model_output, review = \
+                                self.train_step(self.model, example, device[0])
 
-                    # Release pytorch object to reduce memory footprint
-                    del example  # likely to be numpy
-                    del model_output
-                    del review
+                        for hook in hooks:
+                            hook.post_step(self, example, model_output, review)
 
-                if self.iteration == 0 and epoch_start:
-                    # epoch_start is used to prevent raising an error,
-                    # when train_iterator has only one element
-                    raise Exception(
-                        f'Got an empty train iterator: {train_iterator}'
-                    )
+                        # Release pytorch object to reduce memory footprint
+                        del example
+                        del model_output
+                        del review
+                            
+                        with self.train_timer['time_per_backward']:
+                            loss.backward(retain_graph=False)
+                        del loss
 
-                # Fix for next loop
-                self.iteration += 1
+                    else:
+                        with self.train_timer['time_per_step']:
+                            # The data parallel idea here follows the idea from
+                            # torch.nn.parallel.DataParallel.
+                            # We also use the same functions
+                            # (i.e. replicate, parallel_apply and gather).
+                            #
+                            # The difference is, that we need no scatter,
+                            # because we simply use multiple examples and
+                            # the gather must only be applied on the loss.
+
+                            # Move copies of the model to each GPU
+                            replicas = replicate(self.model, device[:len(example)])
```

Comment:
should we not do this before the virtual minibatch loop because now we are copying the same model to the devices in each iteration

Reply:
I copied the pattern from torch.nn.parallel.DataParallel.
They do the replicate in each forward call.
I haven't measured the overhead.
To keep the code simple, I would keep it for the moment here.

Reply:
I think it makes sense to copy it in each forward call if you also always do a backwards step because the model is changing, but in our case we are just doing a backward step once for multiple forwards so we could copy it once

Reply:
I will keep this point for another PR.